### DISCUSSION TITLE
Running Avalonia11 Template on Linux

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#9735] [Q&A]
I'm trying to get the new Avalonia11 templates running on Linux. Desktop works, but...

For `Avalonia11.Web`, I get this

```
App: /usr/share/dotnet/packs/Microsoft.NET.Runtime.WebAssembly.Sdk/7.0.1/WasmAppHost/WasmAppHost.dll
Architecture: x64
Framework: 'Microsoft.AspNetCore.App', version '7.0.0' (x64)
.NET location: /usr/share/dotnet/

No frameworks were found.
```

For `Avalonia11.Android`, I had to uninstall `jdk` package and install `jdk11-openjdk`. I now get this

```
  Xamarin.Android.Tooling.targets(70, 5): [XARSD7000] System.ComponentModel.Win32Exception (2): An error occurred trying to start process '/usr/bin/java' with working directory '/home/hanuman/GitHub/HanumanInstitute.MvvmDialogs/samples/Avalonia11/Avalonia11.Android'. No such file or directory
```

I'm just trying to get the basic template to run...

Version: 11-preview4
OS: Garuda Linux (Arch-based)

**Comment by kekekeks** [comment]
Your system is missing ASP.NET Core shared runtime and /usr/bin/java binary. Make sure that those are installed.

**Reply by mysteryx93** [reply]
I have installed
- dotnet-host
- dotnet-runtime
- dotnet-runtime-6.0
- dotnet-sdk
- dotnet-sdk-6.0
- dotnet-targeting-pack
- dotnet-targeting-pack-6.0

Basically every package in pacman that contains "dotnet".

As for `/usr/bin/java`, it is a symlink pointing to jdk19. Recreated the symlink and now that works.

`sudo ln -s /usr/lib/jvm/java-11-openjdk/bin/java /usr/bin/java`

**Comment by mysteryx93** [comment]
OK finally got the Android application running on Linux! BUT *very* very slowly. Did a bit of reading, and some say that KVM can be used to accelerate it? I'm already using KVM to run Windows in a VM on my system, but don't know if my kernel and security settings are right to allow Android Emulator to use it.

Anyone has details of how to ensure KVM is used with the Android Emulator in Jetbrains Rider in Linux? It's quite a specific set of tools.

Also, by default, the template runs on API 33, does it also support older OS, and how do I configure that?

Still remains to get the Web version running.

**Reply by kekekeks** [reply]
You can try running android-x86 in KVM with virgl enabled. It's certainly faster than the emulated ARM one.

**Reply by mysteryx93** [reply]
would I manually configure KVM in Virtual Manager the same way I did for Windows, or still use Rider to start the emulator? If the emulator is external, how will Rider bind to it to send the application?

**Reply by kekekeks** [reply]
You can use `adb connect` via TCP and Rider would recognize android-x86 VM as a physical device.

**Reply by mysteryx93** [reply]
Found the Android-x86 instructions [here in 2.a](https://www.android-x86.org/documentation/debug.html)

Is there a way to configure this in the Virtual Manager instead of starting it from the command-line every time?
```
kvm -soundhw es1370 -net nic -net user,hostfwd=tcp::4444-:5555 -cdrom buildname.iso
```

And... `kvm` isn't a valid command in Arch, it's started with `virsh`

Found instructions to [expose USB debugging here](https://linuxconfig.org/how-to-use-bridged-networking-with-libvirt-and-kvm), but it's incredibly complicated, I get stuck at `sudo ip link set enp0s29u1u1 up` where I'm not exactly sure what to do, it doesn't have instructions for arch, and it may be outdated.

There has to be a simpler way?

**Reply by kekekeks** [reply]
I told `virt-manager` that I want a "Debian testing" virtual machine and then enabled 3D acceleration in options. That was sufficient for Android x86 to run with VirGL. Then it's just configuring network in Android and using `adb connect` with VM's IP address from the host.
